### ISSUE TITLE
Feature Request: Allow Specification of Proxmox Version for Installation

### ISSUE THREAD

**Issue by thystips** [lae/ansible-role-proxmox#256]
Hello,

First of all, thank you for maintaining this project, it's been extremely helpful in automating my deployments.

I would like to request a feature that allows users to specify the exact version of Proxmox to install during the setup process. In my specific use case, I deploy tools that are only compatible with certain minor versions of Proxmox. Therefore, being able to define the Proxmox version is crucial to ensure compatibility and avoid potential issues.

Currently, the role installs the latest available version, which may lead to incompatibilities with certain tools or scripts that haven't been updated to work with newer releases. By enabling version specification, users can ensure that their deployments are stable and consistent across different environments.

Thank you for considering this feature.

**Comment by lae** [comment]
This is likely going to be very tricky, I think. I can't think up off the top of my head how one would grab a specific "minor" version from the apt repositories for all relevant packages. If you have any information regarding how to go about this, if upstream PVE even makes it possible (e.g. if old releases are not pruned), that'd be helpful.

**Comment by thystips** [comment]
After checking the PVE [repository](http://download.proxmox.com/debian/pve/dists/bookworm/pve-no-subscription/binary-amd64/), it appears that multiple versions are available for each package. This suggests that it should be possible to specify a particular version to install with Ansible.

As a workaround for my specific use case, I could use a custom repository with tailored packages, but this approach is less generalizable. I believe allowing version specification directly in the role would be a more flexible solution.

**Comment by lae** [comment]
Yeah but like, my concern is that there isn't a way to pin *all* packages to an associated minor version. You can install a specific version of a single package, in our case `proxmox-ve`[^pkgs], but what about its dependencies? For example `proxmox-ve` 8.2.0 depends on `pve-qemu-kvm` whose latest is 9.0.2. Pulling proxmox-ve=8.1.0 is still gonna pull pve-qemu-kvm=9.0.2, which sounds like it might cause issues?

[^pkgs]: https://github.com/lae/ansible-role-proxmox/blob/develop/tasks/identify_needed_packages.yml#L5

**Comment by thystips** [comment]
You're right, version need to be set on `proxmox-ve``, `pve-kernel` and `pve-qemu-kvm` but if a strict version is set for `proxmox-ve` in `_pve_install_packages`, `pve_extra_packages` can be used for other packages without add complexity in role.
Is it a better way ?

**Comment by lae** [comment]
> Pulling proxmox-ve=8.1.0 is still gonna pull pve-qemu-kvm=9.0.2, which sounds like it might cause issues?

I gave a very bad example. `pve-qemu-kvm` should totally be fine if on a newer version.

The real issue is that `proxmox-ve` is just a meta package and pinning it effectively does nothing to prevent any dependencies from being upgraded:

```
$ apt-cache show proxmox-ve=9.0.0
Package: proxmox-ve
Architecture: all
Version: 9.0.0
Priority: optional
Section: admin
Maintainer: Proxmox Support Team <support@proxmox.com>
Installed-Size: 25
Provides: proxmox-virtual-environment
Depends: apt, openssh-client, openssh-server, proxmox-archive-keyring, proxmox-default-kernel, proxmox-kernel-helper (>= 9.0~), pve-manager (>= 9.0~), pve-qemu-kvm, qemu-server (>= 9.0~), systemd-sysv
Filename: dists/trixie/pve-no-subscription/binary-amd64/proxmox-ve_9.0.0_all.deb
Size: 5136
MD5sum: e3cdac3288b17bcf5e54f94737f271f9
SHA256: 6ef4c76183473c509114225ad11dc7e9ea493d3999ba187f0f9a682c5a5a3344
SHA512: 7ce3c792a814c08daa3aeee9dd909cc0f3e10f24896e4fd4e0b1ca65b5ba2d4c82d631e646d345d61182c95accbbf3c5aea43230dee87d95dee30123dff2bd92
Homepage: https://www.proxmox.com
Description: Proxmox Virtual Environment
 The Proxmox Virtual Environment is an easy to use Open Source
 virtualization platform for running Virtual Appliances and Virtual
 Machines. This is a meta package which will install everything
 needed.
Description-md5:
```

If I am making the right assumptions, the only viable version pinning is on MAJOR.MINOR, e.g. 9.0 or 9.1. In Bookworm, for example, `proxmox-ve=8.0.2` is the latest 8.0 that exists, but `pve-cluster` goes up to 8.0.8; in other words, given the loose version constraints (which is good) on `proxmox-ve`/`pve-manager`/etc, there's no reliable way to go any more specific.

We could script against `dpkg-query` (or maybe `python-apt` directly, I haven't looked at its API) to scan our list of staged packages (`_pve_install_packages`) and their dependencies to install for the latest patch version of a minor release. It should be doable. But I think it'll end up being brittle and a maintenance burden for not much value. Like, `pve-manager=9.0.0~8` has a version constraint on `pve-docs>8.2.4`—but it has 9.x versions! And `pve-firewall` has no version constraints, but the latest version is 6.0.4? How do I *actually* identify the right package versions in a generic manner without reviewing each dependency in isolation? I think it might be a little fun to implement, though.

That said, I don't think it's a good idea to be pinning PVE anyway outside of major versions (i.e. the Debian release). Staff suggest the same here: https://forum.proxmox.com/threads/hold-the-version-of-proxmox-ve.124985/. The scripts and tools that expect a specific minor version should really use looser version constraints. In semver, minor version updates are supposed to remain backward compatible, so developers of those should be operating on that premise in the first place. It would be very uncommon that a minor version breaks something for them so self-imposed crutches like that are a waste of time. Unless of course they never planned to maintain what they write in the first place.

Anyway, given that this feature is more likely to end up breaking setups than not, I'm gonna vote to close this issue.

**Comment by thystips** [comment]
You're right, this can be closed — and the linked article covers exactly that.

For more specific needs where [Proxmox Offline Mirror](https://pom.proxmox.com/) might not be sufficient, an alternative I've been using is [aptly](https://www.aptly.info/) and its snapshot feature, combined with the `pve_repository` variable to point to the mirror. It gives a bit more flexibility for pinning the repository state at a specific point in time.

**Comment by lae** [comment]
yeah, +1 for aptly. I recall using it with this role to provide an on-site mirror during the earlier days of this project actually...
